packages/firestore/src/api/user_data_writer.ts

```diff
@@ -17,99 +17,112 @@
 
 import * as firestore from '@firebase/firestore-types';
 
-import {
-  ArrayValue,
-  BlobValue,
-  FieldValue,
-  ObjectValue,
-  RefValue,
-  ServerTimestampValue,
-  TimestampValue
-} from '../model/field_value';
+import * as api from '../protos/firestore_proto_api';
+
 import { DocumentReference, Firestore } from './database';
-import * as log from '../util/log';
 import { Blob } from './blob';
+import { GeoPoint } from './geo_point';
 import { Timestamp } from './timestamp';
+import { DocumentKey } from '../model/document_key';
+import { normalizeByteString, normalizeTimestamp } from '../model/values';
+import {
+  getLocalWriteTime,
+  getPreviousValue,
+  isServerTimestamp
+} from '../model/server_timestamps';
+import { fail } from '../util/assert';
+import { forEach } from '../util/obj';
 
 export type ServerTimestampBehavior = 'estimate' | 'previous' | 'none';
 
 /**
  * Converts Firestore's internal types to the JavaScript types that we expose
  * to the user.
  */
-export class UserDataWriter<T> {
+export class UserDataWriter<T = firestore.DocumentData> {
   constructor(
     private readonly firestore: Firestore,
     private readonly timestampsInSnapshots: boolean,
     private readonly serverTimestampBehavior?: ServerTimestampBehavior,
     private readonly converter?: firestore.FirestoreDataConverter<T>
   ) {}
 
-  convertValue(value: FieldValue): unknown {
-    if (value instanceof ObjectValue) {
-      return this.convertObject(value);
-    } else if (value instanceof ArrayValue) {
-      return this.convertArray(value);
-    } else if (value instanceof RefValue) {
-      return this.convertReference(value);
-    } else if (value instanceof BlobValue) {
-      return new Blob(value.internalValue);
-    } else if (value instanceof TimestampValue) {
-      return this.convertTimestamp(value.value());
-    } else if (value instanceof ServerTimestampValue) {
-      return this.convertServerTimestamp(value);
+  convertValue(value: api.Value): unknown {
+    if ('nullValue' in value) {
+      return null;
+    } else if ('booleanValue' in value) {
+      return value.booleanValue!;
+    } else if ('integerValue' in value) {
+      return Number(value.integerValue!);
```

Comment:
This seems like it should call `normalizeNumber` rather than just assuming it knows how to do this.